Client/Frontend/Library/BookmarkDetailPanel.swift

```diff
@@ -0,0 +1,338 @@
+        self.bookmarkNodeGUID = bookmarkNodeGUID
+        self.bookmarkNodeType = bookmarkNodeType
+        self.parentBookmarkFolder = parentBookmarkFolder
+
+        super.init(profile: profile)
+
+        self.tableView.accessibilityIdentifier = "Bookmark Detail"
+        self.tableView.keyboardDismissMode = .onDrag
+        self.tableView.register(TextFieldTableViewCell.self, forCellReuseIdentifier: BookmarkDetailFieldCellIdentifier)
+        self.tableView.register(OneLineTableViewCell.self, forCellReuseIdentifier: BookmarkDetailFolderCellIdentifier)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, closure: { _ in
+            self.navigationController?.popViewController(animated: true)
+        })
+
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .save, closure: { _ in
+            self.save().uponQueue(.main) { _ in
+                self.navigationController?.popViewController(animated: true)
+            }
+        })
+    }
+
+    override func applyTheme() {
+        super.applyTheme()
+
+        if let current = navigationController?.visibleViewController as? Themeable, current !== self {
+            current.applyTheme()
+        }
+
+        tableView.backgroundColor = UIColor.theme.tableView.headerBackground
+    }
+
+    override func reloadData() {
+        profile.places.getBookmarksTree(rootGUID: BookmarkRoots.RootGUID, recursive: true).uponQueue(.main) { result in
+            guard let rootFolder = result.successValue as? BookmarkFolder else {
+                // TODO: Handle error case?
+                self.bookmarkFolders = []
+                self.tableView.reloadData()
+                return
+            }
+
+            var bookmarkFolders: [(folder: BookmarkFolder, indent: Int)] = []
+
+            func addFolder(_ folder: BookmarkFolder, indent: Int = 0) {
+                // Do not append the top "root" folder to this list as
+                // bookmarks cannot be stored directly within it.
+                if folder.guid != BookmarkRoots.RootGUID {
+                    bookmarkFolders.append((folder, indent))
+                }
+
+                for case let childFolder as BookmarkFolder in folder.children ?? [] {
+                    // Any "root" folders (i.e. "Mobile Bookmarks") should
+                    // have an indentation of 0.
+                    if childFolder.isRoot {
+                        addFolder(childFolder)
+                    }
+                    // Otherwise, all non-root folder should increase the
+                    // indentation by 1.
+                    else {
+                        addFolder(childFolder, indent: indent + 1)
```

Comment:
What is the max indent level? This should be capped, at which point it stops indenting (which is ok for those outliers that have deeply nested folders, they can still find a way to use this).

Reply:
Added a commit to programmatically determine an appropriate max indent level depending on screen width.